Add api docs

jhrcek opened the pull request:
I started working on the api docs, but before I invest more time in this, I'd like to get some initial feedback. Is this anywhere near to what you expected the api docs to look like? Do you prefer more prose or more code examples?

jhrcek commented on src/Routing/Duplex.purs:
```diff
@@ -128,6 +167,16 @@ as f g (RouteDuplex enc dec) = RouteDuplex (enc <<< f) (Parser.as identity g dec
 int :: RouteDuplex' String -> RouteDuplex' Int
 int = as show Parser.int
 
+-- | Make Boolean codec out of a String codec.
```
It's hard to come up with a good descriptions of what those functions of type `RouteDuplex' a -> RouteDuplex' b` do without talking about printing and parsing separately :P

natefaubion replied:
Maybe "restricts" or "refines"?

> Refines a codec of Strings to Booleans, where `true` and `false` are the strings "true" and "false".

What do you think?

natefaubion commented on src/Routing/Duplex.purs:
```diff
@@ -64,9 +71,12 @@ instance applicativeRouteDuplex :: Applicative (RouteDuplex i) where
   dimap f g (RouteDuplex enc dec) = RouteDuplex (f >>> enc) (g <$> dec)
 
+-- | Use given codec to parse a value of type `o` out of String
+-- | (typically representing part of URL) or produce a `RouteError` if the parsing fails.
```
I think we could clarify what "part of a URL" means. The `String` is always expected to be the path, query parameters, and hash.

natefaubion commented on src/Routing/Duplex.purs:
```diff
@@ -64,9 +71,12 @@ instance applicativeRouteDuplex :: Applicative (RouteDuplex i) where
 parse :: forall i o. RouteDuplex i o -> String -> Either Parser.RouteError o
 parse (RouteDuplex _ dec) = Parser.run dec
 
+-- | Use given codec to transform a value of type `i` to a String.
```
Maybe instead of "transform to a String", we can say that it prints the `i` type to a URL path, query parameters, and hash. Kind of like with `parse`, since it's just the inverse.

natefaubion commented on src/Routing/Duplex.purs:
```diff
@@ -79,12 +89,41 @@ suffix (RouteDuplex enc dec) s = RouteDuplex (\a -> enc a <> Printer.put s) (dec
 path :: forall a b. String -> RouteDuplex a b -> RouteDuplex a b
 path = flip (foldr prefix) <<< String.split (Pattern "/")
 
+-- | Modifies given `codec` by requiring it to be prefixed with '/'.
```
I don't think that we need to put `codec` in backticks because it's not referencing any code anywhere.

> Modifies a codec so that it must be prefixed with "/".

natefaubion commented on src/Routing/Duplex.purs:
```diff
@@ -79,12 +89,41 @@ suffix (RouteDuplex enc dec) s = RouteDuplex (\a -> enc a <> Printer.put s) (dec
 root :: forall a b. RouteDuplex a b -> RouteDuplex a b
 root = path ""
 
+-- | Modifies given `codec` by requiring it to consume the entire rest of input.
```
Same here.
> Modifies a codec so that it must consume the rest of the input, otherwise it fails.

natefaubion commented on src/Routing/Duplex.purs:
```diff
@@ -79,12 +89,41 @@ suffix (RouteDuplex enc dec) s = RouteDuplex (\a -> enc a <> Printer.put s) (dec
 end :: forall a b. RouteDuplex a b -> RouteDuplex a b
 end (RouteDuplex enc dec) = RouteDuplex enc (dec <* Parser.end)
 
+-- | Process path segment as a String.
```
I don't think "process" is the right word.

> Consumes and prints a single path segment.

What do you think?

jhrcek replied:
It's kind of hard because of RouteDuplex being double purpose thing. `Consumes and prints`  sounds more specific than just process, but since you can never use a codec for both purposes at a given call site, I'd replace `and` with `or`.

jhrcek replied:
I went for "Consumes or prints a single path segment." in the end.

natefaubion commented on src/Routing/Duplex.purs:
```diff
@@ -79,12 +89,41 @@ suffix (RouteDuplex enc dec) s = RouteDuplex (\a -> enc a <> Printer.put s) (dec
+-- | parse segment "abc"       == Right "abc"
+-- | parse segment "abc%20def" == Right "abc def" -- automatic decoding of uri components
+-- | parse segment "abc/def"   == Right "abc"
+-- | parse segment "/abc"      == Right "" -- the empty string before the first '/'
```
I like the examples! Maybe we could add an one here that uses `root`.

jhrcek replied:
From library user's perspective I always find examples useful. I'll add one with root as you suggest.
But in general I feel slightly uncomfortable without verifying all of them with some kind of "doctest" tool. Does purescript have something along those lines? I'll definitely check all the examples in the repl, but for long term maintenance having an automated way to check the examples would be better. I found https://github.com/hdgarrood/purs-doctest/ but that seems to be abandoned..

jhrcek commented:
I addressed most of your comments and added some more + some unit tests verifying most of the docs examples I added. I'll still add some more tomorrow and let you know when it's ready for review.

jhrcek commented on src/Routing/Duplex.purs:
```diff
@@ -64,40 +72,116 @@ instance applicativeRouteDuplex :: Applicative (RouteDuplex i) where
 param p = RouteDuplex (Printer.param p) (Parser.param p)
 
+-- TODO not sure what's use case for this combinator? Is it like part of path that can optionally be present? Or is it about query params?
 flag :: RouteDuplex' String -> RouteDuplex' Boolean
```
Just one question: It's not clear to me what would be the use case of this combinator..

natefaubion replied:
This is for query parameters like "?flag", where it's presence indicates true, and it's absence indicates false. Contrast with `boolean (param "flag")` which requires it to be `?flag=true` or `?flag=false`, and absence fails to parse.

jhrcek commented:
I consider it ready for 2nd round of review. Most of the functions in `Routing.Duplex` are documented except for `record` and `prop` (I don't understand PS records that well yet to be able to document it properly).

jhrcek commented on src/Routing/Duplex.purs:
```diff
@@ -106,31 +200,101 @@ many1 :: forall f a b.
 boolean :: RouteDuplex' String -> RouteDuplex' Boolean
 boolean = as show Parser.boolean
 
+-- | This does nothing (internally it's defined as identity).
```
This comment for `string` sounds dumb, but I couldn't think of better description :P

natefaubion replied:
This mainly exists to work around https://github.com/natefaubion/purescript-routing-duplex/issues/6

jhrcek commented:
Hello @natefaubion just a friendly reminder to check this PR again. I think I addressed your comments and I'd like to avoid this PR to rot.

natefaubion commented:
Thanks for all the work, I made a few edits for typos and consistency.
